### PR TITLE
feat(VInput): auto hiding details if no messages/counter

### LINKED ISSUE #7434: [Feature Request] Hide details on inputs unless there are details to show
### Problem to solve
Vuetify inputs have space reserved under them for showing errors and hints. This adds to the problem of large forms having too much white space as mentioned in #1846. Using `hide-details` will remove this extra space, but it also hides any hint text or validation errors. I want to remove that space unless there is a hint or error that needs to be displayed.

### Proposed solution
A new prop, maybe `conditional-details` or `auto-details`, that will hide the white space unless a hint or error is showing.

### PR DESCRIPTION
**Pull request by jacekkarczmarczyk**
## Motivation and Context
closes #7434

## How Has This Been Tested?
playground, jest

## Markup:

<details>

```vue
<template>
  <v-app id="inspire">
    <v-container>
      <v-text-field outlined label="Hint" v-model="hint" />
      <v-checkbox label="Rules" v-model="rules" />
      <v-checkbox label="Persistent hint" v-model="persistentHint" />
      <v-divider />
      <v-text-field filled :hint="hint" :persistent-hint="persistentHint" :rules="rules ? [v => v == 'ok' || 'Type &quot;ok&quot;'] : []" label="hide-details = 'auto'" hide-details="auto" />
      <v-text-field filled :hint="hint" :persistent-hint="persistentHint" :rules="rules ? [v => v == 'ok' || 'Type &quot;ok&quot;'] : []" label="hide-details = false" />
      <v-text-field filled :hint="hint" :persistent-hint="persistentHint" :rules="rules ? [v => v == 'ok' || 'Type &quot;ok&quot;'] : []" label="hide-details = true" hide-details />
    </v-container>
  </v-app>
</template>

<script>
export default {
  data: () => ({
    hint: '',
    rules: true,
    persistentHint: true,
  })
}
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any feature but make things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and breaking changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/components/VInput/VInput.ts [hunk 1]
```diff
@@ -20,6 +20,7 @@ import {
 // Types
 import { VNode, VNodeData, PropType } from 'vue'
 import mixins from '../../util/mixins'
+import { PropValidator } from 'vue/types/options'
```
Comment (KaelWD) [inline packages/vuetify/src/components/VInput/VInput.ts]: `PropType` is already up there :point_up:

**Comment by johnleider** [comment]
Should this check rules as well? The only issue I can see is having an input without messages next to one without. It would create a weird style issue.

**Comment by mrksoftware** [comment]
Hi guys, any news on this? For us would be really useful! 

I’ve started to implement something similar on each input field manually with the new grid system in Vuetify 2 and I have to say that it looks great!

I also didn’t experienced any weird issue with layout, when one field has error in one row, the vertical space needed for the error message is maintained across all the line. (I can provide screens if you need)

On the other hand, without something like this the UI looks not ok cause the horizontal space between fields is quite different from the vertical space between fields.

I’d like to know your opinion, thanks in advance :)

**Comment by jacekkarczmarczyk** [comment]
> Should this check rules as well? 

What do you mean? It shouldn't care about rules but about messages (which are there if rule has failed validation)

> The only issue I can see is having an input without messages next to one without. It would create a weird style issue.

Then this feature (or layout where 2 fields are next to each other) simply should not be used

@johnleider

File: packages/vuetify/src/components/VInput/VInput.ts [hunk 2]
```diff
@@ -66,7 +66,7 @@ export default baseMixins.extend<options>().extend({
     classes (): object {
       return {
         'v-input--has-state': this.hasState,
-        'v-input--hide-details': this.hideDetails,
+        'v-input--hide-details': this.genMessages() === null,
```
Comment (KaelWD) [inline packages/vuetify/src/components/VInput/VInput.ts]: Not too stoked about using a render method here.

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VInput/VInput.ts]: Fixed

**Pull request merged**